### ISSUE TITLE
cli `ConfigBuilder` is complex, and breaks encapsulation

### ISSUE THREAD

**Issue by danieleades** [foresterre/cargo-msrv#353]
the cli's `ConfigBuilder` is complex, and a little 'inside out'. Also it's not a 'builder'.

the way it works currently is,

1. `CargoMsrvOpts` is parsed from the user's input
2. the options are passed in turn to a set of 'configurators'
3. each configurator knows the internal structure of the options struct
4. each configurator sets the appropriate value by calling a method on the builder

- this is quite hard to follow
- the 'configurators' shouldn't need to 'know' the internal structure of the options (breaks encapsulation)
- this isn't a 'builder' api at all. The 'builder' *does* use a fluent api (where methods take `self`, but it might as well be a plain struct.

The solution I would propose is to move the logic for parsing the 'options' into the options struct itself.

**Comment by danieleades** [comment]
you can see an example of what i mean in #355 (there's a fair bit more to do)

**Comment by foresterre** [comment]
Thanks for the feedback and considerations.

I believe [ConfigBuilder](https://github.com/foresterre/cargo-msrv/blob/ede4adaa76691bda7b2e4927afc5f09669390343/src/config.rs#L327-L419) is a builder. It follows a style which is described by the Rust book as a [consuming builder](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html#consuming-builders:). 

Notably, the 'configurators' themselves are not part of the builder. They're only used to configure it, by handling some of the more complex [cases](https://github.com/foresterre/cargo-msrv/blob/ede4adaa76691bda7b2e4927afc5f09669390343/src/cli/configurators/custom_check.rs) where changing the CLI would break backwards compatibility, and I wasn't yet willing to do so.

Yes, the configurators aren't strictly necessary, I could also written some plain functions, instead of boxing them in per input, or kept it like it was before: one huge function which configured everything. Something needs to call the builder and its methods after all.

> each configurator knows the internal structure of the options struct
This is probably indeed not necessary, but sometimes practicality wins from purity =). One way to fix the encapsulation problem you describe would be by added getter methods on the Opts struct, and removing the `pub(in cli::)`. However, it's a rather marginal win, since the visibility is already scoped to the `cli` module, and adds a bunch of getter methods which would only be used to configure the builder again. Encapsulation can be important, but I don't think it's truly an issue here.


`ConfigBuilder` is certainly getting more complex :). I've noted as much when I last worked on this part. Below is an excerpt from the source:
https://github.com/foresterre/cargo-msrv/blob/ede4adaa76691bda7b2e4927afc5f09669390343/src/config.rs#L203-L206

While it's getting lengthy, and doesn't integrate very well with some ideas I've been working on recently (layered configuration), I don't think it's too complicated; in my opinion it's rather straightforward.

> The solution I would propose is to move the logic for parsing the 'options' into the options struct itself.

I'm not necessarily against this, but something to consider is the following: the more subcommands, options, and flags are added, the more methods will be added to `CargoMsrvOpts`. I think that moving all cli 'parsing' methods to `CargoMsrvOpts` does just that, _move_ the logic to some other place.
